8359105: RISC-V: No need for acquire fence in safepoint poll during JNI calls

Anjian-Wen opened the pull request:
Acquire fence removal in safepoint_poll

At least in jdk11, when comes to safepoint::end, it will invoke SafepointMechanism::disarm_local_poll to change the polling_word_offset, which may cause a race when thread come to visit polling_word_offset in native_trans state, so we used to use acquire fence.

Since the disarm_local_poll has been removed from SafepointSynchronize::end, Thread disarm itself in 
JavaThread::check_special_condition_for_native_trans when trans from native. it seems that there is no need for acquire fence in safepoint_poll.

[0] https://github.com/openjdk/jdk/pull/20420

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8359105](https://bugs.openjdk.org/browse/JDK-8359105): RISC-V: No need for acquire fence in safepoint poll during JNI calls (**Enhancement** - P4)


### Reviewers
 * [Robbin Ehn](https://openjdk.org/census#rehn) (@robehn - **Reviewer**)
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)
 * [Feilong Jiang](https://openjdk.org/census#fjiang) (@feilongjiang - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/25709/head:pull/25709` \
`$ git checkout pull/25709`

Update a local copy of the PR: \
`$ git checkout pull/25709` \
`$ git pull https://git.openjdk.org/jdk.git pull/25709/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 25709`

View PR using the GUI difftool: \
`$ git pr show -t 25709`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/25709.diff">https://git.openjdk.org/jdk/pull/25709.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/25709#issuecomment-2961884698)
</details>

Anjian-Wen commented:
@robehn Hi, Can you help to review the patch?

robehn commented:
Hey, sure.

The description says "backport of", as you are changing master I think that description is wrong.

The downcallLinker_riscv.cpp is the same case as the native transition.
So I don't see why you would keep that acquire ?

AFIACT no one should use acquire, thus this should then mean that we can remove "bool acquire" argument from safepoint_poll().

That arm still have acquire in their downlinker seems like an oversight?
8337657 only have one reviewer, I think it should have been cought there.
(please note https://wiki.openjdk.org/display/HotSpot/Pushing+a+HotSpot+change, two reviewers required)
@dchuyko can you open a new issue and look the acquire in downcall linker for aarch64 ?

Thanks, Robbin

dchuyko commented:
> @dchuyko can you open a new issue and look at the acquire in downcall linker for aarch64 ?

Thanks for the reminder, created JDK-8359252. It was intentional to limit the scope of the original change to JNI (due the amount of testing and usages).

Anjian-Wen commented:
Thanks for your review!! I have changed the description. 
About the safepoint_poll, I think it makes sence that there is no need to keep "bool acquire" argument when we change the last 'true' in downcallLinker_riscv.cpp, and I have updated the patch.

> Hey, sure.
> 
> The description says "backport of", as you are changing master I think that description is wrong.
> 
> The downcallLinker_riscv.cpp is the same case as the native transition. So I don't see why you would keep that acquire ?
> 
> AFIACT no one should use acquire, thus this should then mean that we can remove "bool acquire" argument from safepoint_poll().
> 
> That arm still have acquire in their downlinker seems like an oversight? 8337657 only have one reviewer, I think it should have been cought there. (please note https://wiki.openjdk.org/display/HotSpot/Pushing+a+HotSpot+change, two reviewers required) @dchuyko can you open a new issue and look at the acquire in downcall linker for aarch64 ?
> 
> Thanks, Robbin

robehn commented:
> > @dchuyko can you open a new issue and look at the acquire in downcall linker for aarch64 ?
> 
> Thanks for the reminder, created JDK-8359252. It was intentional to limit the scope of the original change to JNI (due the amount of testing and usages).

Thank you!

robehn reviewed (approved):
Thanks, looks good!

feilongjiang reviewed (approved):
Looks fine. Thanks!

Anjian-Wen commented:
/integrate

Anjian-Wen commented:
@robehn @feilongjiang @RealFYang  Thanks for your review and approve !

feilongjiang commented:
/sponsor
